pkg/resources/mutation/cluster/provisioningcluster.go

```diff
@@ -53,10 +97,257 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
+// updates the cluster's spec to mount the secret to the control plane nodes, and configures kube-apisever to use the admission configuration file;
+// If a PSACT is unset in the cluster, handlePSACT does the cleanup for both the secret and cluster's spec.
+func (m *ProvisioningClusterMutator) handlePSACT(request *admission.Request, cluster *v1.Cluster) (*admissionv1.AdmissionResponse, error) {
+	if cluster.Name == "local" || cluster.Spec.RKEConfig == nil {
+		return admission.ResponseAllowed(), nil
+	}
+	parsedVersion, err := psa.GetClusterVersion(cluster.Spec.KubernetesVersion)
+	if err != nil {
+		return nil, fmt.Errorf("[provsioning cluster mutator] failed to parse cluster version: %w", err)
+	}
+	if parsedRangeLessThan123(parsedVersion) {
+		return admission.ResponseAllowed(), nil
+	}
+
+	secretName := fmt.Sprintf(SecretName, cluster.Name)
+	mountPath := fmt.Sprintf(MountPath, GetRuntime(cluster.Spec.KubernetesVersion))
+	templateName := cluster.Spec.DefaultPodSecurityAdmissionConfigurationTemplateName
+
+	switch request.Operation {
+	case admissionv1.Delete:
+		err := m.secret.Delete(cluster.Namespace, secretName, &metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("[provsioning cluster mutator] failed to delete the secret: %w", err)
+		}
+	case admissionv1.Create, admissionv1.Update:
+		if cluster.DeletionTimestamp != nil {
+			return admission.ResponseAllowed(), nil
+		}
+		if templateName == "" {
+			err := m.secret.Delete(cluster.Namespace, secretName, &metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to delete the secret: %w", err)
+			}
+			// drop relevant fields if they exist in the cluster
+			cluster = DropMachineSelectorFile(MachineSelectorFileForPSA(secretName, mountPath), cluster)
+			args := GetKubeAPIServerArg(cluster)
+			if args[KubeAPIAdmissionConfigOption] == mountPath {
+				delete(args, KubeAPIAdmissionConfigOption)
+				cluster = SetKubeAPIServerArg(args, cluster)
+			}
+		} else {
+			// Now, handle the case of PSACT being set when creating or updating the cluster
+			template, err := m.psact.Get(templateName)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return admission.ResponseBadRequest(err.Error()), nil
+				}
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to get psact: %w", err)
+			}
+			fileContent, err := psa.GenerateAdmissionConfigFile(template, cluster.Spec.KubernetesVersion)
+			if err != nil {
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to generate admission configuration from PSACT [%s]: %w", templateName, err)
+			}
+			anno := map[string]string{
+				SecretAnnotation: cluster.Name,
+			}
+			data := map[string][]byte{
+				SecretKey: fileContent,
+			}
+			_, err = m.createOrUpdateSecret(cluster.Namespace, secretName, data, anno)
```

Comment:
nit: If we do not need the variable we should not return it since this is the only call to our private function.

Reply:
fixed